design/authenticators/application-identity/application-identity.md

```diff
@@ -0,0 +1,466 @@
+Identities are related across authenticators, and how they are different.
+ 
+For the sake of simplicity, from here and throughout the document, Application 
+Identities will be defined only in the host's annotations, and not in the ID. 
+In future authenticators we will not develop an option to define it in the ID and it
+is available in the K8s Authenticator only for backwards compatibility.
+
+The Application Identity validation can be roughly split into 2 parts, where
+ the second part can also be split into 3 parts:
+ - Extract the Application Identity from the host
+ - Validate the Application Identity is defined correctly
+ - Validate the Application Identity matches the authentication request
+
+Let's break down each part to better understand them.
+
+### Defining Application Identity Validation
+
+#### Extract the Application Identity from the host
+
+Let's look into the Azure Authenticator's Application Identity class:
+```
+class ApplicationIdentity
+
+    def initialize(role_annotations:, service_id:, logger:)
+        @role_annotations = role_annotations
+        @service_id       = service_id
+        @logger           = logger
+    end
+    
+    def constraints
+        @constraints ||= {
+          subscription_id:          constraint_value("subscription-id"),
+          resource_group:           constraint_value("resource-group"),
+          user_assigned_identity:   constraint_value("user-assigned-identity"),
+          system_assigned_identity: constraint_value("system-assigned-identity")
+        }.compact
+    end
+    
+    private
+    
+    # check the `service-id` specific constraint first to be more granular
+    def constraint_value constraint_name
+        annotation_value("authn-azure/#{@service_id}/#{constraint_name}") ||
+          annotation_value("authn-azure/#{constraint_name}")
+    end
+    
+    def annotation_value name
+        annotation = @role_annotations.find { |a| a.values[:name] == name }
+        
+        # return the value of the annotation if it exists, nil otherwise
+        if annotation
+          @logger.debug(LogMessages::Authentication::RetrievedAnnotationValue.new(name))
+          annotation[:value]
+    end
+end
+```
+
+The class gets the host's annotations and the authenticator's service-id, 
+and then builds a `constraints` hash from the annotations. 
+
+This is the case also in [`authn-k8s`](/app/domain/authentication/authn_k8s/application_identity.rb), 
+with only one difference as the K8s Authenticator also looks in the host's ID. 
+
+Putting that aside we can see that the behavior is exactly the same. 
+In this case the general class will be defined as follows:
+- input
+  - authenticator name
+  - authenticator service-id
+  - annotations
+  - List of relevant annotations (e.g `[subscription-id, resource-group
+  , system-assigned-identity, user-assigned-identity]
+  ` for `authn
+  -azure`)
+- output
+  - `constraints` hash.
```

Comment:
@jonahx is there a better option than a hash? Please keep this question in mind while reading the rest of the doc and answer it once you're finished.